### PR TITLE
Fix issues with multiple identical sources (#2811)

### PR DESCRIPTION
**Pull request by hbenl**

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -9,7 +9,10 @@ const test = url.searchParams.get("test");
 
 // During testing, make sure we clear local storage before importing
 // any code that might instantiate preferences from local storage.
-if (test) {
+// If the url contains a "navigated" parameter, we assume this is the
+// second part of a test that contains a navigation and don't clear
+// local storage in that case.
+if (test && !url.searchParams.get("navigated")) {
```
Comment (loganfsmyth) [inline src/main.js:15]: Would it be feasible to have the test harness script itself do this before each test? It seems like that would make things simpler.

**Review by loganfsmyth — COMMENTED** [review]
Since I'm not super familiar with the state of things, this is a bit tough to review. Could you describe what overall architecture you're aiming for with this patch? 

Say a recording has 4 copies of a source, and they all have the same URL, and the first 2 have one set of contents, and the second 2 have another set of contents, what is the expected output? Is that something you're trying to address, or is this still assuming that all sources with the same URL are identical, for instance? Is the goal the present each source as a separate entry in the source-tree, or are we still presenting a single entry in the source tree to represent many sources?

Separately, are there changes to the protocol that we could make that would make any of this easier? It seems like all of the code around this is quite complex.

**Comment by hbenl** [comment]
Due to sourcemapping we can't assume that all sources with the same URL are identical, of course. I'm relying entirely on the existing method `ThreadFront._chooseSourceIdList()` to group sources that "belong together" (i.e. they are original/generated/pretty-printed variants of the same "real" source from a single source load event) and pick a preferred and optionally an alternate source from each group.
I assume that the preferred (or alternate) sources for the same URL have the same content (although I don't verify this anywhere), so I pick the first preferred/alternate source for each URL and try to use that in the source tree and the editor (both of them never had a way to properly deal with multiple sources for the same URL anyway).
When running an analysis, I add the locations for all corresponding sources (the corresponding sources of a preferred/alternate source are the preferred/alternate sources with the same URL) to get all hits.
When adding breakpoints from a previous session, I only add them for one of the corresponding sources (but when the analysis is run, the other corresponding sources will be added to the locations).
In the QuickOpenModal I only add one source per URL.
See also #2570, #2600 and #2810.

File: src/main.js (hunk 1 above)
Comment (hbenl) [reply src/main.js:15]: Yeah, that would be a lot cleaner. We should refactor how we're initialising the frontend anyway, so I'll have another look at this then.

**Review by loganfsmyth — APPROVED** [review]
Alright, gonna approve, though I admit I'm not super confident about it because I don't feel like I have a good sense of the frontend codebase at this point.

**Pull request merged**